pkg/admin/autorecovery/restore.go

```diff
@@ -0,0 +1,180 @@
+)
+
+type Manager struct {
+	storage    data.StoragePath
+	saveFailed bool
+}
+
+func NewManager(storage data.StoragePath, saveFailed bool) (*Manager, error) {
+	if storage == "" {
+		return nil, fmt.Errorf("`storage` argument is empty")
+	}
+
+	return &Manager{storage: storage, saveFailed: saveFailed}, nil
+}
+
+func (m *Manager) PerformRestore() error {
+	if err := storageShouldExist(m.storage); err != nil {
+		return err
+	}
+
+	existingState, err := GetState(m.storage)
+	if err != nil {
+		return err
+	}
+
+	restoreCandidate, err := getCandidateForRestore(m.storage, existingState)
+	if err != nil {
+		return err
+	}
+
+	/*
+		Preparations - initial file system operations:
+		  COPY    $STORAGE/$CANDIDATE -> /var/lib/microshift.tmp
+		  COPY    /var/lib/microshift -> $STORAGE/failed/$DATE_DEPLOY-ID.tmp
+		  CREATE  $STORAGE/state.json.new
+		  -       $STORAGE/$PREVIOUSLY_RESTORED									Should already exist, nothing to do.
+
+		"Closing the transaction"
+		  RENAME    /var/lib/microshift.tmp -> /var/lib/microshift
+						This operation is first, because it's the most important one - whole point of the restore is to get to running state.
+		  RENAME    $STORAGE/failed/DATE_DEPLOY-ID.tmp -> $STORAGE/failed/DATE_DEPLOY-ID
+		  RENAME    $STORAGE/state.json.new -> $STORAGE/state.json
+		  RENAME    $STORAGE/$PREVIOUSLY_RESTORED -> $STORAGE/restored/$PREVIOUSLY_RESTORED
+	*/
+
+	// Copies/creations into intermediate destinations
+	var oldData *data.AtomicDirCopy
+	if m.saveFailed {
+		oldDataBackupName, err := GetBackupName()
+		if err != nil {
+			return err
+		}
+		failedStorage := m.storage.SubStorage(failedSubstorageName)
+		if err := os.MkdirAll(string(failedStorage), 0600); err != nil {
+			return fmt.Errorf("failed to create %q subdirectory: %w", failedSubstorageName, err)
+		}
+		oldData = &data.AtomicDirCopy{Source: config.DataDir, Destination: failedStorage.GetBackupPath(oldDataBackupName)}
+		if err := oldData.CopyToIntermediate(); err != nil {
+			return fmt.Errorf("old microshift data: %w", err)
+		}
+	}
+
+	newData := data.AtomicDirCopy{Source: m.storage.GetBackupPath(restoreCandidate.Name()), Destination: config.DataDir}
+	if err := newData.CopyToIntermediate(); err != nil {
+		if rollbackErr := oldData.RollbackIntermediate(); rollbackErr != nil {
+			klog.ErrorS(rollbackErr, "Failed to rollback intermediate state for old data")
+		}
+		return fmt.Errorf("new microshift data: %w", err)
+	}
+
+	newState := NewState(m.storage, restoreCandidate.Name())
+	if err := newState.SaveToIntermediate(); err != nil {
+		if rollbackErr := oldData.RollbackIntermediate(); rollbackErr != nil {
+			klog.ErrorS(rollbackErr, "Failed to rollback intermediate state for old data")
+		}
+		if rollbackErr := newData.RollbackIntermediate(); rollbackErr != nil {
+			klog.ErrorS(rollbackErr, "Failed to rollback intermediate state for new data")
+		}
+		return fmt.Errorf("new state file: %w", err)
+	}
+
+	// Renames into final destinations
```

Comment:
If any of the following fail the rollbacks are performed within their own functions, right? What about the other variables (newData, newState, oldData)? Would a defer help with the data about what to rollback help?

Reply:
When this fails in the middle of renaming, I don't think there's much to do...
For example, if `oldData.RenameToFinal()` fails, then to roll everything back we'd have to move `oldData.Intermediate` -> `/var/lib/microshift` (because it already contains the data we wanted to restore). I feel like this would be very volatile twister game.
If the FS would offer transactions, we'd be golden, but I think renaming microshift data into place first is good enough (although we still return error and rc=1 so it might not be interpreted that way, maybe we need more exit codes?)